wrappers/c/zxing.cpp

```diff
@@ -0,0 +1,202 @@
+	return (zxing_ImageFormat) cppformat;
+}
+
+/*
+ * ...
+ */
+
+/*
+ * ZXing/BarcodeFormat.h
+ */
+
+zxing_BarcodeFormat zxing_BarcodeFormatFromString (const char * format)
+{
+    return (zxing_BarcodeFormat) ZXing::BarcodeFormatFromString(std::string(format));
+}
+
+// TODO:
+//zxing_BarcodeFormat zxing_BarcodeFormatsFromString (const char * formats)
+//{
+//    return (zxing_BarcodeFormat) ZXing::BarcodeFormatsFromString(std::string(formats));
+//}
+
+/*
+ * ZXing/DecodeHints.h
+ */
+
+zxing_DecodeHints zxing_DecodeHints_new ()
+{
+	ZXing::DecodeHints * cpphints = new ZXing::DecodeHints();
+	return (zxing_DecodeHints) cpphints;
+}
+
+void zxing_DecodeHints_free (zxing_DecodeHints hints)
+{
+	ZXing::DecodeHints * cpphints = (ZXing::DecodeHints *) hints;
+	delete cpphints;
+}
+
+void zxing_DecodeHints_setTryHarder (zxing_DecodeHints hints, bool tryHarder)
+{
+	ZXing::DecodeHints * cpphints = (ZXing::DecodeHints *) hints;
+	cpphints->setTryHarder(tryHarder);
+}
+
+void zxing_DecodeHints_setTryDownscale (zxing_DecodeHints hints, bool tryDownscale)
+{
+	ZXing::DecodeHints * cpphints = (ZXing::DecodeHints *) hints;
+	cpphints->setTryDownscale(tryDownscale);
+}
+
+void zxing_DecodeHints_setFormats (zxing_DecodeHints hints, zxing_BarcodeFormat formats)
+{
+	ZXing::DecodeHints * cpphints = (ZXing::DecodeHints *) hints;
+	ZXing::BarcodeFormat cppformats = (ZXing::BarcodeFormat) formats;
+	cpphints->setFormats(cppformats);
+}
+
+void zxing_DecodeHints_setBinarizer (zxing_DecodeHints hints, zxing_Binarizer binarizer)
+{
+	ZXing::DecodeHints * cpphints = (ZXing::DecodeHints *) hints;
+	ZXing::Binarizer cppbinarizer = (ZXing::Binarizer) binarizer;
+	cpphints->setBinarizer(cppbinarizer);
+}
+
+void zxing_DecodeHints_setEanAddOnSymbol (zxing_DecodeHints hints, zxing_EanAddOnSymbol eanAddOnSymbol)
+{
+	ZXing::DecodeHints * cpphints = (ZXing::DecodeHints *) hints;
+	ZXing::EanAddOnSymbol cppeanAddOnSymbol = (ZXing::EanAddOnSymbol) eanAddOnSymbol;
+	cpphints->setEanAddOnSymbol(cppeanAddOnSymbol);
+}
+
+void zxing_DecodeHints_setTextMode (zxing_DecodeHints hints, zxing_TextMode textMode)
+{
+	ZXing::DecodeHints * cpphints = (ZXing::DecodeHints *) hints;
+	ZXing::TextMode cpptextMode = (ZXing::TextMode) textMode;
+	cpphints->setTextMode(cpptextMode);
+}
+
+/*
+ * ...
+ */
+
+/*
+ * ZXing/Result.h
+ */
+
+void zxing_Result_free (zxing_Result result)
+{
+    ZXing::Result * cppresult = (ZXing::Result *) result;
+    delete cppresult;
+}
+
+bool zxing_Result_isValid (zxing_Result result)
+{
+	ZXing::Result * cppresult = (ZXing::Result *) result;
+	return cppresult->isValid();
+}
+
+zxing_BarcodeFormat zxing_Result_format (zxing_Result result)
+{
+	ZXing::Result * cppresult = (ZXing::Result *) result;
+	return (zxing_BarcodeFormat) cppresult->format();
+}
+
+/*
+ * ...
+ */
+
+/*
+ * ZXing/ReadBarcode.h
+ */
+
+zxing_Result zxing_ReadBarcode (const zxing_ImageView iv, const zxing_DecodeHints hints)
+{
+	const ZXing::ImageView & cppiv = (const ZXing::ImageView &) iv;
```

Comment:
`zxing_ImageView` is a pointer, not a reference. This might have something todo with your crashes?

Reply:
Don't know. However, initially I had this:

```cpp
zxing_Result zxing_ReadBarcode (const zxing_ImageView iv, const zxing_DecodeHints hints)
{
	const ZXing::ImageView * cppiv = (const ZXing::ImageView *) iv;
	const ZXing::DecodeHints * cpphints = (const ZXing::DecodeHints *) hints;
	ZXing::Result cppresult = ZXing::ReadBarcode(cppiv, cpphints);
	// ...
```

With that version the C++ compiler complains:

```
zxing.cpp: In function ‘ZXing::Result* zxing_ReadBarcode(zxing_ImageView, zxing_DecodeHints)’:
zxing.cpp:177:54: error: invalid initialization of reference of type ‘const ZXing::ImageView&’ from expression of type ‘const ZXing::ImageView*’
  177 |         ZXing::Result cppresult = ZXing::ReadBarcode(cppiv, cpphints);
      |                                                      ^~~~~
In file included from zxing.cpp:7:
/home/siiky/.local/include/ZXing/ReadBarcode.h:21:37: note: in passing argument 1 of ‘ZXing::Result ZXing::ReadBarcode(const ImageView&, const DecodeHints&)’
   21 | Result ReadBarcode(const ImageView& buffer, const DecodeHints& hints = {});
      |                    ~~~~~~~~~~~~~~~~~^~~~~~
```

Reply:
You need to dereference the two pointers: `ZXing::ReadBarcode(*cppiv, *cpphints)`

Reply:
Oh! That was it, yeah, thanks.

Reply:
This may have really been the cause of the segfault. Though I still can't scan a simple QR code, the example program doesn't segfault anymore.